src/lib/array.jl

```diff
@@ -165,10 +167,21 @@ end
 # This is also used by comprehensions, which do guarantee iteration order.
 # Not done for pmap, presumably because all is lost if you are relying on its order.
 _tryreverse(m, backs, Δ) = backs, Δ
-_tryreverse(m::typeof(map), backs, Δ) = reverse(backs), reverse(Δ)
+_tryreverse(m::typeof(map), backs, Δ) = _reverse(backs), _reverse(Δ)
 
 _tryreverse(m, x) = x
-_tryreverse(m::typeof(map), x) = reverse(x)
+_tryreverse(m::typeof(map), x) = _reverse(x)
+
+# Fallback
+_reverse(x) = reverse(x)
+
+# Known cases in the standard library on which `reverse` errors (issue #1393)
+_reverse(x::LowerTriangular) = UpperTriangular(reverse(parent(x)))
+_reverse(x::UpperTriangular) = LowerTriangular(reverse(parent(x)))
+_reverse(x::UnitLowerTriangular) = UnitUpperTriangular(reverse(parent(x)))
+_reverse(x::UnitUpperTriangular) = UnitLowerTriangular(reverse(parent(x)))
```

Comment:
Should they maybe call `_reverse` instead of `reverse`?
```suggestion
_reverse(x::LowerTriangular) = UpperTriangular(_reverse(parent(x)))
_reverse(x::UpperTriangular) = LowerTriangular(_reverse(parent(x)))
_reverse(x::UnitLowerTriangular) = UnitUpperTriangular(_reverse(parent(x)))
_reverse(x::UnitUpperTriangular) = UnitLowerTriangular(_reverse(parent(x)))
```

Reply:
I think for this selection of types it does not matter, but when extending this one may run into infinite recursion (I can't think of an example right now).

Reply:
As long as all implementations of `_reverse` guarantee that the recurse only into an array/structure whose number of nested levels is smaller, it seems infinite recursion is not possible? And the definitions here and below seem to satisfy this property: the array on the LHS is a wrapper of the array on the RHS that `_reverse` is called with, so the number of levels is reduced by 1.

Reply:
Yeah you are right, one just needs to be careful when extending this in the future.